### PR TITLE
[CPDLP-3450] NPQ Contracts migration from ECF to NPQ

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net.mcas.ms/browse/CPDLP-3450

### Changes proposed in this pull request

* Created `Migration::Ecf::NpqContract` model
* Created `Migration::Migrators::Contract` migrator
* Created tests and factories.

### CODE REVIEW

File: app/services/migration/migrators/contract.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    def call
+      migrate(self.class.ecf_contracts) do |ecf_contract|
+        course = courses_by_identifier[ecf_contract.course_identifier]
+
+        ecf_statements(ecf_contract).find_each do |ecf_statement|
+          statement = ::Statement.find_by!(ecf_id: ecf_statement.id)
+
+          contract_template = ::ContractTemplate.find_or_create_by!(ecf_id: ecf_contract.id)
```
Comment (cwrw) [inline app/services/migration/migrators/contract.rb:28]:
```suggestion
          contract_template = ::ContractTemplate. find_or_initialize_by(ecf_id: ecf_contract.id)
```

File: spec/services/migration/migrators/contract_spec.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      create(
+        :ecf_migration_npq_contract,
+        npq_lead_provider:,
+        cohort:,
+        version: "0.0.1",
+      )
+    end
+
+    describe "#call" do
+      it "sets the created Contract and ContractTemplate attributes correctly" do
```
Comment (cwrw) [inline spec/services/migration/migrators/contract_spec.rb:71]: might be worth using the shared context here for migrators

File: app/services/migration/migrators/contract.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    def call
+      migrate(self.class.ecf_contracts) do |ecf_contract|
+        course = courses_by_identifier[ecf_contract.course_identifier]
+
+        ecf_statements(ecf_contract).find_each do |ecf_statement|
+          statement = ::Statement.find_by!(ecf_id: ecf_statement.id)
+
+          contract_template = ::ContractTemplate.find_or_initialize_by(ecf_id: ecf_contract.id)
```
Comment (ethax-ross) [inline app/services/migration/migrators/contract.rb:28]: As the attributes have the same names could do:


```
    SHARED_ATTRIBUTES = %w[
      created_at
      updated_at
      service_fee_percentage
      output_payment_percentage
      per_participant
      number_of_payment_periods
      recruitment_target
      service_fee_installments
      targeted_delivery_funding_per_participant
      monthly_service_fee
      special_course
    ].freeze

contract_template.update!(ecf_contract.attributes.slice(SHARED_ATTRIBUTES))
```

Or if there are only a few we don't want to migrate across, something like:

```
contract_template.update!(ecf_contract.attributes.excluding(:id))
```

File: spec/services/migration/migrators/contract_spec.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      )
+      create(
+        :contract,
+        statement:,
+        course:,
+        contract_template:,
+      )
+    end
+
+    def setup_failure_state
```
Comment (ethax-ross) [inline spec/services/migration/migrators/contract_spec.rb:51]: Can we add a comment around why this should be a failure?

File: spec/services/migration/migrators/contract_spec.rb (hunk 2 above)
Comment (ethax-ross) [reply spec/services/migration/migrators/contract_spec.rb:71]: I think it already is?

File: spec/services/migration/migrators/contract_spec.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        expect(contracts.count).to eq(1)
+
+        contract = statement.contracts.first
+        expect(contract.course.identifier).to eq(ecf_resource1.course_identifier)
+
+        contract_template = contract.contract_template
+        expect(contract_template.ecf_id).to eq(ecf_resource1.id)
+
+        attrs = ecf_resource1.attributes.slice(
+          :service_fee_percentage,
```
Comment (ethax-ross) [inline spec/services/migration/migrators/contract_spec.rb:95]: I think the ECF NPQContract factory and the NPQ ContractTemplate factory both have the same values for each of these anyway, which means the test will pass regardless of if they get set in the migrator 😬 can we change one of the factories to be slightly different so its testing the attribute assignment?

**Review by ethax-ross — APPROVED** [review]
Looks good, just one thing I noticed about the factories happen to have the same attribute values in ECF/NPQ; if we make them different the test will be more robust, but other than that I think this is good to go 👍

File: spec/services/migration/migrators/contract_spec.rb (hunk 5 above)
Comment (mooktakim) [reply spec/services/migration/migrators/contract_spec.rb:95]: Makes sense, updating now

**Review by ethax-ross — APPROVED** [review]

**Review by leandroalemao — COMMENTED** [review]
Missed that one but well done @mooktakim tks

**Pull request merged**